pkg/build/cmd/verifystarlark.go

```diff
@@ -0,0 +1,153 @@
+	"path/filepath"
+	"strings"
+
+	"github.com/urfave/cli/v2"
+)
+
+func mapSlice[I any, O any](a []I, f func(I) O) []O {
+	o := make([]O, len(a))
+	for i, e := range a {
+		o[i] = f(e)
+	}
+	return o
+}
+
+// VerifyStarlark is the CLI Action for verifying Starlark files in a workspace.
+// It expects a single context argument which is the path to the workspace.
+// The actual verification procedure can return multiple errors which are
+// joined together to be one holistic error for the action.
+func VerifyStarlark(c *cli.Context) error {
+	if c.NArg() != 1 {
+		var message string
+		if c.NArg() == 0 {
+			message = "ERROR: missing required argument <workspace path>"
+		}
+		if c.NArg() > 1 {
+			message = "ERROR: too many arguments"
+		}
+
+		if err := cli.ShowSubcommandHelp(c); err != nil {
+			return cli.Exit(err.Error(), 1)
+		}
+		return cli.Exit(message, 1)
+	}
+
+	workspace := c.Args().Get(0)
+	verificationErrs, executionErr := verifyStarlark(c.Context, workspace, buildifierLintCommand)
+	if executionErr != nil {
+		return cli.Exit(executionErr.Error(), 1)
+	}
+
+	if len(verificationErrs) == 0 {
+		return nil
+	}
+
+	noun := "file"
+	if len(verificationErrs) > 1 {
+		noun += "s"
+	}
+	return fmt.Errorf("verification failed for %d %s:\n%s",
+		len(verificationErrs),
+		noun,
+		strings.Join(
+			mapSlice(verificationErrs, func(e error) string { return e.Error() }),
+			"\n",
+		))
+}
+
+type commandFunc = func(path string) (command string, args []string)
+
+func buildifierLintCommand(path string) (string, []string) {
+	return "buildifier", []string{"-lint", "warn", path}
+}
+
+// verifyStarlark walks all directories starting at provided workspace path and
+// verifies any Starlark files it finds.
+// Starlark files are assumed to end with the .star extension.
+// The verification relies on linting frovided by the 'buildifier' binary which
+// must be in the PATH.
+// A slice of verification errors are returned, one for each file that failed verification.
+// If any execution of the `buildifier` command fails, this is returned separately.
+// commandFn is executed on every Starlark file to determine the command and arguments to be executed.
+// The caller is trusted and it is the callers responsibility to ensure that the resulting command is safe to execute.
+func verifyStarlark(ctx context.Context, workspace string, commandFn commandFunc) ([]error, error) {
+	var verificationErrs []error
+	var executionErr error
+
+	// All errors from filepath.WalkDir are filtered by the fs.WalkDirFunc.
+	// The anonymous function used here never returns an error.
+	// Lstat or ReadDir errors are reported as verificationErrors.
+	// If any execution of the `buildifier` command fails, it is reported as executionErr and
+	// any verification of subsequent files is skipped.
+	if err := filepath.WalkDir(workspace, func(path string, d fs.DirEntry, err error) error {
+		// Skip verification of the file or files within the directory if there is an error
```

Comment:
Please also check inside the WalkFunc if the context has been cancelled and return the context's error if so.

Reply:
I'm also not very familiar with using the `context` package. How should one correctly check the channel returned by `context.Done()` in a WalkFunc such as this?

Reply:
You normally do it somehow like this:

```go
select {
  case <- ctx.Done():
    return ctx.Err()
  default:
}
```

In the ctx.Done case the err will be most likely the reason why the context has been cancelled, so you could either take this one as executionErr or use some other error to just get out of the walk callback 😄

Reply:
Just added a test case to check the context and realized that we already handle context cancellation automatically when I pass the context to each command execution in `cmd := exec.CommandContext(ctx, command, args...)`. Do you still think it is worthwhile to check the context before running any command?

Reply:
I don't think that's necessary as any longrunning step would notice it anyway and abort anyway 🙂